rama-net/src/address/socket_address.rs

```diff
@@ -0,0 +1,257 @@
+        Self::try_from(bytes.as_slice())
+    }
+}
+
+impl TryFrom<&[u8]> for SocketAddress {
+    type Error = OpaqueError;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
+        let s = std::str::from_utf8(bytes).context("parse authority from bytes")?;
+        s.try_into()
+    }
+}
+
+impl serde::Serialize for SocketAddress {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let address = self.to_string();
+        address.serialize(serializer)
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for SocketAddress {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+        s.try_into().map_err(serde::de::Error::custom)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn assert_eq(s: &str, sock_address: SocketAddress, ip_addr: &str, port: u16) {
+        assert_eq!(
+            sock_address.ip_addr().to_string(),
+            ip_addr,
+            "parsing: {}",
+            s
+        );
+        assert_eq!(sock_address.port(), port, "parsing: {}", s);
+    }
+
+    #[test]
+    fn test_parse_valid() {
+        for (s, (expected_ip_addr, expected_port)) in [
+            ("[::1]:80", ("::1", 80)),
+            ("127.0.0.1:80", ("127.0.0.1", 80)),
+            (
+                "[2001:db8:3333:4444:5555:6666:7777:8888]:80",
+                ("2001:db8:3333:4444:5555:6666:7777:8888", 80),
+            ),
+        ] {
+            let msg = format!("parsing '{}'", s);
+
+            assert_eq(s, s.parse().expect(&msg), expected_ip_addr, expected_port);
+            assert_eq(
+                s,
+                s.try_into().expect(&msg),
+                expected_ip_addr,
+                expected_port,
+            );
+            assert_eq(
+                s,
+                s.to_owned().try_into().expect(&msg),
+                expected_ip_addr,
+                expected_port,
+            );
+            assert_eq(
+                s,
+                s.as_bytes().try_into().expect(&msg),
+                expected_ip_addr,
+                expected_port,
+            );
+            assert_eq(
+                s,
+                s.as_bytes().to_vec().try_into().expect(&msg),
+                expected_ip_addr,
+                expected_port,
+            );
+        }
+    }
+
+    #[test]
+    fn test_parse_invalid() {
+        for s in [
+            "",
+            "-",
+            ".",
+            ":",
+            ":80",
+            "-.",
+            ".-",
+            "::1",
+            "127.0.0.1",
+            "[::1]",
+            "2001:db8:3333:4444:5555:6666:7777:8888",
+            "[2001:db8:3333:4444:5555:6666:7777:8888]",
+            "example.com",
+            "example.com:",
+            "example.com:-1",
+            "example.com:999999",
```

Comment:
```suggestion
            "example.com:999999",
            "example.com:80",
```

Make sure to also add a valid Authority, as that's valid for an authority but shouldn't be valid for a SocketAddress